src/Solver/time_stepping/GhostTimeClusterWithCopy.cpp

```diff
@@ -0,0 +1,199 @@
+    receiveRegionsStates[region] = RequiresMpiTesting;
+  }
+
+  duplicatedCopyRegions.resize(numberOfRegions);
+  duplicatedGhostRegions.resize(numberOfRegions);
+  for (size_t region = 0; region < numberOfRegions; ++region) {
+    const size_t copyRegionSize = meshStructure->copyRegionSizes[region] * sizeof(real);
+    const size_t ghostRegionSize = meshStructure->ghostRegionSizes[region] * sizeof(real);
+    if constexpr (CommType == MPI::DataTransferMode::CopyInCopyOutDevice) {
+      duplicatedCopyRegions[region] = static_cast<real*>(device.api->allocGlobMem(copyRegionSize));
+      duplicatedGhostRegions[region] =
+          static_cast<real*>(device.api->allocGlobMem(ghostRegionSize));
+    } else {
+      duplicatedCopyRegions[region] =
+          static_cast<real*>(device.api->allocPinnedMem(copyRegionSize));
+      duplicatedGhostRegions[region] =
+          static_cast<real*>(device.api->allocPinnedMem(ghostRegionSize));
+    }
+  }
+}
+
+template <MPI::DataTransferMode CommType>
+GhostTimeClusterWithCopy<CommType>::~GhostTimeClusterWithCopy() {
+  for (size_t region = 0; region < numberOfRegions; ++region) {
+    device.api->destroyGenericStream(prefetchCopyRegionsStreams[region]);
+    device.api->destroyGenericStream(prefetchGhostRegionsStreams[region]);
+    if constexpr (CommType == MPI::DataTransferMode::CopyInCopyOutDevice) {
+      device.api->freeMem(duplicatedCopyRegions[region]);
+      device.api->freeMem(duplicatedGhostRegions[region]);
+    } else {
+      device.api->freePinnedMem(duplicatedCopyRegions[region]);
+      device.api->freePinnedMem(duplicatedGhostRegions[region]);
+    }
+  }
+}
+
+template <MPI::DataTransferMode CommType>
+void GhostTimeClusterWithCopy<CommType>::sendCopyLayer() {
+  SCOREP_USER_REGION("sendCopyLayer", SCOREP_USER_REGION_TYPE_FUNCTION)
+  assert(ct.correctionTime > lastSendTime);
+
+  lastSendTime = ct.correctionTime;
+  auto prefetchedRegions = prefetchCopyLayer();
+
+  while (not prefetchedRegions.empty()) {
+    for (auto region = prefetchedRegions.begin(); region != prefetchedRegions.end();) {
+      auto* stream = prefetchCopyRegionsStreams[*region];
+      if (device.api->isStreamWorkDone(stream)) {
+        MPI_Isend(duplicatedCopyRegions[*region],
+                  static_cast<int>(meshStructure->copyRegionSizes[*region]),
+                  MPI_C_REAL,
+                  meshStructure->neighboringClusters[*region][0],
+                  timeData + meshStructure->sendIdentifiers[*region],
+                  seissol::MPI::mpi.comm(),
+                  meshStructure->sendRequests + (*region));
+        sendQueue.push_back(*region);
+        region = prefetchedRegions.erase(region);
+      } else {
+        ++region;
+      }
+    }
+  }
+}
+
+template <MPI::DataTransferMode CommType>
+void GhostTimeClusterWithCopy<CommType>::receiveGhostLayer() {
+  SCOREP_USER_REGION("receiveGhostLayer", SCOREP_USER_REGION_TYPE_FUNCTION)
+  assert(ct.predictionTime > lastSendTime);
```

Comment:
You can change this to >=, there's an open issue somewhere that this assert is buggy

Reply:
Ok. I changed in both places, namely: here and in the parent class.